addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/net/BasicSocketSession.java

```diff
@@ -0,0 +1,415 @@
+     *
+     * @see
+     * org.openhab.binding.russound.internal.net.SocketSession#addListener(org.openhab.binding.russound.internal.net.
+     * SocketSessionListener)
+     */
+    @Override
+    public void addListener(SocketSessionListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener cannot be null");
+        }
+        _listeners.add(listener);
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#clearListeners()
+     */
+    @Override
+    public void clearListeners() {
+        _listeners.clear();
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see
+     * org.openhab.binding.russound.internal.net.SocketSession#removeListener(org.openhab.binding.russound.internal.net.
+     * SocketSessionListener)
+     */
+    @Override
+    public boolean removeListener(SocketSessionListener listener) {
+        return _listeners.remove(listener);
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#connect()
+     */
+    @Override
+    public void connect() throws IOException {
+        disconnect();
+
+        _client = new Socket(_host, _port);
+        _client.setKeepAlive(true);
+        _client.setSoTimeout(1000); // allow reader to check to see if it should stop every 1 second
+
+        _logger.debug("Connecting to {}:{}", _host, _port);
+        _writer = new PrintStream(_client.getOutputStream());
+        _reader = new BufferedReader(new InputStreamReader(_client.getInputStream()));
+
+        new Thread(_dispatcher).start();
+        new Thread(_responseReader).start();
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#disconnect()
+     */
+    @Override
+    public void disconnect() throws IOException {
+        if (isConnected()) {
+            _logger.debug("Disconnecting from {}:{}", _host, _port);
+
+            _dispatcher.stopRunning();
+            _responseReader.stopRunning();
+
+            _writer.close();
+            _writer = null;
+
+            _reader.close();
+            _reader = null;
+
+            _client.close();
+            _client = null;
+
+            _responses.clear();
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#isConnected()
+     */
+    @Override
+    public boolean isConnected() {
+        return _client != null && _client.isConnected();
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#sendCommand(java.lang.String)
+     */
+    @Override
+    public synchronized void sendCommand(String command) throws IOException {
+        if (command == null) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        // if (command.trim().length() == 0) {
```

Comment:
That whole class will be removed very shortly (I've been testing the SocketChannelSession implementation and it appears pretty stable).